flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/parser/ClientParser.java

```diff
@@ -0,0 +1,192 @@
+import static org.apache.flink.sql.parser.impl.FlinkSqlParserImplConstants.SEMICOLON;
+import static org.apache.flink.sql.parser.impl.FlinkSqlParserImplConstants.SET;
+import static org.apache.flink.sql.parser.impl.FlinkSqlParserImplConstants.SHOW;
+import static org.apache.flink.sql.parser.impl.FlinkSqlParserImplConstants.STATEMENT;
+
+/**
+ * ClientParser use {@link FlinkSqlParserImplTokenManager} to do lexical analysis. It cannot
+ * recognize special hive keywords yet because Hive has a slightly different vocabulary compared to
+ * Flink's, which causes the ClientParser misunderstanding some Hive's keywords to IDENTIFIER. But
+ * the ClientParser is only responsible to check whether the statement is completed or not and only
+ * cares about a few statements. So it's acceptable to tolerate the inaccuracy here.
+ */
+public class ClientParser implements SqlCommandParser {
+
+    /** A dumb implementation. TODO: remove this after unifying the SqlMultiLineParser. */
+    @Override
+    public Optional<Operation> parseCommand(String command) {
+        return Optional.empty();
+    }
+
+    public Optional<StatementType> parseStatement(String statement) throws SqlExecutionException {
+
+        return getStatementType(new TokenIterator(statement.trim()));
+    }
+
+    private static class TokenIterator implements Iterator<Token> {
+
+        private final FlinkSqlParserImplTokenManager tokenManager;
+        private Token currentToken;
+
+        public TokenIterator(String statement) {
+            tokenManager =
+                    new FlinkSqlParserImplTokenManager(
+                            new SimpleCharStream(new StringReader(statement)));
+            // means to switch to "BACK QUOTED IDENTIFIER" state to support '`xxx`' in Flink SQL
+            tokenManager.SwitchTo(2);
+            this.currentToken = tokenManager.getNextToken();
+            // this will make all tokens link together
+            checkIncompleteStatement();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return currentToken.kind != EOF;
+        }
+
+        @Override
+        public Token next() {
+            Token before = currentToken;
+            currentToken = before.next;
+            return before;
+        }
+
+        boolean nextTokenMatched(int kind) {
+            return hasNext() && next().kind == kind;
+        }
+
+        /**
+         * check these follow special cases: 1. "". 2. "COMPILE/EXECUTE STATEMENT SET BEGIN\n INSERT
+         * xxx;". 3. "EXPLAIN COMPILE/EXECUTE STATEMENT SET BEGIN\n INSERT xxx;". 4. "EXPLAIN BEGIN
+         * STATEMENT SET;". 5. string not ended with ';'.
+         */
+        void checkIncompleteStatement() {
+            // case 1
+            if (currentToken.kind == EOF) {
+                continueReadInput();
+            }
+
+            Token head = currentToken, tail = tokenManager.getNextToken();
+
+            // case 2, 3 and 4
+            boolean setNotEnded =
```

Comment:
The `explain` also supports executing statement set with explain details. I think we need to rethink the behavior here.